server/src/main/java/io/spine/server/delivery/ShardedStream.java

```diff
@@ -0,0 +1,361 @@
+    private final Class<I> targetIdClass;
+    private final Subscriber subscriber;
+    private final Publisher publisher;
+    private final ExternalMessageObserver channelObserver;
+
+    /**
+     * A lazily-intialized converted for the sharded messages.
+     */
+    @Nullable
+    private ShardedMessageConverter<I, M, E> converter;
+
+    ShardedStream(AbstractBuilder<I, E, ?, ? extends ShardedStream> builder) {
+        this.boundedContextName = builder.boundedContextName;
+        this.key = builder.key;
+        this.tag = builder.tag;
+        this.targetIdClass = builder.targetIdClass;
+        final Class<E> envelopeCls = getEnvelopeClass();
+        final ChannelId channelId = toChannelId(builder.boundedContextName, key, envelopeCls);
+        this.subscriber = builder.transportFactory.createSubscriber(channelId);
+        this.publisher = builder.transportFactory.createPublisher(channelId);
+
+        this.channelObserver = new ExternalMessageObserver(builder.consumer);
+        this.subscriber.addObserver(channelObserver);
+    }
+
+    @SuppressWarnings("unchecked")  // Ensured by the generic type definition.
+    private Class<E> getEnvelopeClass() {
+        return (Class<E>) MESSAGE_CLASS.getArgumentIn(getClass());
+    }
+
+    private Class<I> getTargetIdClass() {
+        return targetIdClass;
+    }
+
+    private ShardedMessageConverter<I, M, E> converter() {
+        if(converter == null) {
+            converter = newConverter();
+        }
+        return converter;
+    }
+
+    protected abstract ShardedMessageConverter<I, M, E> newConverter();
+
+    public ShardingKey getKey() {
+        return key;
+    }
+
+    public DeliveryTag<E> getTag() {
+        return tag;
+    }
+
+    public final void post(I targetId, E messageEnvelope) {
+        checkNotNull(messageEnvelope);
+
+        final ShardedMessage shardedMessage = converter().convert(targetId, messageEnvelope);
+        checkNotNull(shardedMessage);
+
+        post(shardedMessage);
+    }
+
+    private void post(ShardedMessage message) {
+        final ExternalMessage externalMessage = ExternalMessages.of(message, boundedContextName);
+        publisher.publish(externalMessage.getId(), externalMessage);
+    }
+
+    @Override
+    public void close() {
+        if (channelObserver != null) {
+            subscriber.removeObserver(channelObserver);
+        }
+    }
+
+    private static <E extends MessageEnvelope<?, ?, ?>>
+    ChannelId toChannelId(BoundedContextName boundedContextName,
+                          ShardingKey key,
+                          Class<E> envelopeClass) {
+        checkNotNull(key);
+        checkNotNull(boundedContextName);
+        checkNotNull(envelopeClass);
+
+        final ClassName className = key.getEntityClass()
+                                       .getClassName();
+        final ShardIndex shardIndex = key.getIndex();
+
+        final String value = on("__").join("bc_", boundedContextName.getValue(),
```

Comment:
This piece looks like a candidate for a utility method of `ChannelIds` class or something other utility that serves the `ChannelId`s. I assume there is a parsing method somewhere, right?

Event if not, having a method that makes string formation clear would be easier to test and understand.

Reply:
Or, just extract as a method.